File: frontend/src/types/task.ts
/**
 * Task types matching backend models
 */

export type TaskType = 'core' | 'bounty';
export type TaskScope = 'mandatory' | 'opt_in' | 'private';
export type TaskDifficulty = 'easy' | 'medium' | 'hard' | 'expert';
export type TaskStatus = 'locked' | 'available' | 'in_progress' | 'submitted' | 'approved' | 'rejected' | 'completed';

export interface Task {
  id: string;
  title: string;
  description: string;
  type: TaskType;
  scope: TaskScope;
  max_participants?: number | null;
  assignee_ids?: string[];
  difficulty: TaskDifficulty;
  min_level_required: number;
  prerequisite_task_ids: string[];
  xp_reward: number;
  skill_tags: string[];
  instructions?: string | null;
  reference_links: string[];
  is_active: boolean;
  order_index: number;
  created_at: string;
  updated_at: string;
  creator_id?: string;
  creator_name?: string | null;
}

export interface UserTask {
  id: string;
  user_id: string;
  user_email?: string | null;
  user_full_name?: string | null;
  task_id: string;
  task: Task;
  status: TaskStatus;
  proof_link?: string | null;
  submission_notes?: string | null;
  submitted_at?: string | null;
  reviewer_id?: string | null;
  mentor_comment?: string | null;
  reviewed_at?: string | null;
  xp_earned: number;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskListResponse {
  tasks: Task[];
  total: number;
  page: number;
  page_size: number;
}

export interface UserTaskListResponse {
  tasks: UserTask[];
  total: number;
  completed_count: number;
  pending_count: number;
  in_progress_count: number;
}

export interface TaskStatsResponse {
  total_tasks: number;
  core_tasks: number;
  bounty_tasks: number;
  active_tasks: number;
  total_xp_available: number;
}

export interface TaskStartRequest {
  task_id: string;
}

export interface TaskSubmitRequest {
  proof_link?: string;
  submission_notes?: string;
}

export interface TaskReviewRequest {
  status: 'approved' | 'rejected';
  mentor_comment?: string;
}

export interface TaskParticipant {
  id: string;
  email: string;
  full_name: string | null;
  avatar_url: string | null;
  role: string;
  level: number;
  current_xp: number;
  task_status: TaskStatus;
  started_at: string | null;
  submitted_at: string | null;
  completed_at: string | null;
  proof_link: string | null;
  mentor_comment: string | null;
  xp_earned: number;
}

export interface TaskDetailResponse {
  id: string;
  title: string;
  description: string;
  type: TaskType;
  scope: TaskScope;
  difficulty: TaskDifficulty;
  min_level_required: number;
  xp_reward: number;
  skill_tags: string[];
  is_active: boolean;
  max_participants: number | null;
  assignee_ids: string[];
  created_at: string;
  updated_at: string;
  total_participants: number;
  in_progress_count: number;
  submitted_count: number;
  completed_count: number;
  rejected_count: number;
  participants: TaskParticipant[];
}

File: backend/app/schemas/task.py
"""
Task Schemas - Request/Response models for task management
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from uuid import UUID
from app.models.task import TaskType, TaskDifficulty, TaskStatus, TaskScope


# ============================================
# Task Schemas
# ============================================

class TaskCreate(BaseModel):
    """Schema for creating a new task (Admin only)"""
    title: str = Field(..., min_length=1, max_length=255)
    description: str
    type: TaskType
    scope: TaskScope = TaskScope.MANDATORY
    difficulty: TaskDifficulty = TaskDifficulty.MEDIUM
    min_level_required: int = 1
    max_participants: Optional[int] = None
    assignee_ids: List[str] = []
    prerequisite_task_ids: List[str] = []
    xp_reward: int = Field(..., ge=0)
    skill_tags: List[str] = []
    instructions: Optional[str] = None
    reference_links: List[str] = []
    is_active: bool = True
    order_index: int = 0


class TaskUpdate(BaseModel):
    """Schema for updating a task (Admin only)"""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    type: Optional[TaskType] = None
    scope: Optional[TaskScope] = None
    difficulty: Optional[TaskDifficulty] = None
    min_level_required: Optional[int] = None
    max_participants: Optional[int] = None
    assignee_ids: Optional[List[str]] = None
    prerequisite_task_ids: Optional[List[str]] = None
    xp_reward: Optional[int] = Field(None, ge=0)
    skill_tags: Optional[List[str]] = None
    instructions: Optional[str] = None
    reference_links: Optional[List[str]] = None
    is_active: Optional[bool] = None
    order_index: Optional[int] = None


class TaskResponse(BaseModel):
    """Public task information"""
    id: UUID
    title: str
    description: str
    type: TaskType
    scope: TaskScope
    difficulty: TaskDifficulty
    min_level_required: int
    max_participants: Optional[int]
    prerequisite_task_ids: List[str]
    xp_reward: int
    skill_tags: List[str]
    instructions: Optional[str]
    reference_links: List[str]
    is_active: bool
    order_index: int
    created_at: datetime
    updated_at: datetime
    creator_id: Optional[UUID]
    creator_name: Optional[str] = None

    class Config:
        from_attributes = True


class TaskListResponse(BaseModel):
    """Paginated task list"""
    tasks: List[TaskResponse]
    total: int
    page: int
    page_size: int


# ============================================
# User Task Schemas
# ============================================

class UserTaskStart(BaseModel):
    """Schema for starting/claiming a task"""
    task_id: UUID


class UserTaskSubmit(BaseModel):
    """Schema for submitting a task for review"""
    proof_link: Optional[str] = Field(None, max_length=500)
    submission_notes: Optional[str] = None

    model_config = {'extra': 'forbid'}


class UserTaskReview(BaseModel):
    """Schema for mentor to review a task"""
    status: TaskStatus = Field(..., description="approved or rejected")
    mentor_comment: Optional[str] = None
    
    class Config:
        use_enum_values = True


class UserTaskResponse(BaseModel):
    """User task progress information"""
    id: UUID
    user_id: UUID
    user_email: Optional[str] = None
    user_full_name: Optional[str] = None
    task_id: UUID
    task: TaskResponse
    status: TaskStatus
    proof_link: Optional[str]
    submission_notes: Optional[str]
    submitted_at: Optional[datetime]
    reviewer_id: Optional[UUID]
    mentor_comment: Optional[str]
    reviewed_at: Optional[datetime]
    xp_earned: int
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class UserTaskListResponse(BaseModel):
    """List of user tasks with status"""
    tasks: List[UserTaskResponse]
    total: int
    completed_count: int
    pending_count: int
    in_progress_count: int


class TaskStatsResponse(BaseModel):
    """Statistics for tasks"""
    total_tasks: int
    core_tasks: int
    bounty_tasks: int
    active_tasks: int
    total_xp_available: int


class TaskParticipant(BaseModel):
    """Task participant information"""
    id: UUID
    email: str
    full_name: Optional[str]
    avatar_url: Optional[str]
    role: str
    level: int
    current_xp: int
    task_status: TaskStatus
    started_at: Optional[datetime]
    submitted_at: Optional[datetime]
    completed_at: Optional[datetime]
    proof_link: Optional[str]
    mentor_comment: Optional[str]
    xp_earned: int

    class Config:
        from_attributes = True


class TaskDetailResponse(BaseModel):
    """Detailed task information with participants"""
    id: UUID
    title: str
    description: str
    type: TaskType
    scope: TaskScope
    difficulty: TaskDifficulty
    min_level_required: int
    xp_reward: int
    skill_tags: List[str]
    is_active: bool
    max_participants: Optional[int]
    assignee_ids: List[str]
    created_at: datetime
    updated_at: datetime
    # Statistics
    total_participants: int
    in_progress_count: int
    submitted_count: int
    completed_count: int
    rejected_count: int
    # Participants list
    participants: List[TaskParticipant]

    class Config:
        from_attributes = True
